builtin/game/chat.lua

```diff
@@ -983,23 +983,60 @@ core.register_chatcommand("days", {
 })
 
 core.register_chatcommand("shutdown", {
-	params = S("[<delay_in_seconds> | -1] [reconnect] [<message>]"),
-	description = S("Shutdown server (-1 cancels a delayed shutdown)"),
+	params = S("[<delay_in_seconds> | -1] [-r] [<message>]"),
+	description = S("Shutdown server (-1 cancels a delayed shutdown, -r allows players to reconnect)"),
 	privs = {server=true},
 	func = function(name, param)
 		local delay, reconnect, message
-		delay, param = param:match("^%s*(%S+)(.*)")
-		if param then
-			reconnect, param = param:match("^%s*(%S+)(.*)")
+		local one, two, three
+		one, two, three = param:match("^(%S+) +(%-r) +(.*)")
+		if one and two and three then
+			-- delay, reconnect and message
+			delay, reconnect, message = one, two, three
+		else
+			-- two arguments
+			one, two = param:match("^(%S+) +(.*)")
+			local done = false
+			if one and two then
+				if tonumber(one) then
+					delay = one
+					if two == "-r" then
+						reconnect = two
+					else
+						message = two
+					end
+					done = true
+				elseif one == "-r" then
+					reconnect, message = one, two
+					done = true
+				end
+			end
+			-- one argument
+			if not done then
+				one = param:match("(.*)")
+				if tonumber(one) then
+					delay = one
+				elseif one == "-r" then
+					reconnect = one
+				else
+					message = one
+				end
+			end
+		end
```

Comment:
Please move this to a function so early return can be used. The logic is quite hard to follow here.